src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -779,13 +779,53 @@ public SolutionCompilationState WithDocumentAttributes<TArg>(
             SolutionState.WithDocumentAttributes(documentId, arg, updateAttributes), documentId);
     }
 
+    /// <summary>
+    /// Creates a new solution instance with the documents specified updated to have the texts specified.
+    /// </summary>
+    /// <remarks>
+    /// This method supports updating ordinary documents and source generated documents.
+    /// </remarks>
     internal SolutionCompilationState WithDocumentTexts(ImmutableArray<(DocumentId documentId, SourceText text)> texts, PreservationMode mode)
-        => UpdateDocumentsInMultipleProjects<DocumentState, SourceText, PreservationMode>(
-            texts,
+    {
+        ImmutableArray<(SourceGeneratedDocumentIdentity, DateTime, SourceText, SyntaxNode?)> sourceGeneratedUpdates = [];
+        var ordinaryDocuments = texts.Any(static d => d.documentId.IsSourceGenerated)
+            ? GetOrdinaryAndSourceGeneratedDocuments(out sourceGeneratedUpdates)
+            : texts;
```

Comment:
could the check just move into GetOrdinaryAndSourceGeneratedDocuments?  i think it would be simpler and not have this weird flow of calling into helpers that assign out params :)

Reply:
i made this changed and pushed to your branch.